decred.go

```diff
@@ -19,110 +21,108 @@ import (
 )
 
 const (
-	// defaultAddressTimeout defines the duration to wait
-	// for new addresses.
+	// defaultAddressTimeout defines the duration to wait for new addresses.
 	defaultAddressTimeout = time.Minute * 10
 
-	// defaultNodeTimeout defines the timeout time waiting for
-	// a response from a node.
+	// defaultNodeTimeout defines the timeout on responses from a node.
 	defaultNodeTimeout = time.Second * 3
 )
 
-var (
-	amgr *Manager
-	wg   sync.WaitGroup
-)
+var amgr *Manager
+
+func testPeer(ctx context.Context, ip net.IP, netParams *chaincfg.Params) {
+	onaddr := make(chan struct{}, 1)
+	verack := make(chan struct{}, 1)
+	config := peer.Config{
+		UserAgentName:    appName,
+		UserAgentVersion: "0.0.1",
+		Net:              netParams.Net,
+		DisableRelayTx:   true,
+
+		Listeners: peer.MessageListeners{
+			OnAddr: func(p *peer.Peer, msg *wire.MsgAddr) {
+				n := make([]net.IP, 0, len(msg.AddrList))
+				for _, addr := range msg.AddrList {
+					n = append(n, addr.IP)
+				}
+				added := amgr.AddAddresses(n)
+				log.Printf("Peer %v sent %v addresses, %d new",
+					p.Addr(), len(msg.AddrList), added)
+				onaddr <- struct{}{}
+			},
+			OnVerAck: func(p *peer.Peer, msg *wire.MsgVerAck) {
+				log.Printf("Adding peer %v with services %v pver %d",
+					p.NA().IP.String(), p.Services(), p.ProtocolVersion())
+				verack <- struct{}{}
+			},
+		},
+	}
 
-func creep(netParams *chaincfg.Params) {
-	defer wg.Done()
+	host := net.JoinHostPort(ip.String(), netParams.DefaultPort)
+	p, err := peer.NewOutboundPeer(&config, host)
+	if err != nil {
+		log.Printf("NewOutboundPeer on %v: %v", host, err)
+		return
+	}
 
-	var wg sync.WaitGroup
+	amgr.Attempt(ip)
+	ctxTimeout, cancel := context.WithTimeout(ctx, defaultNodeTimeout)
+	defer cancel()
+	var dialer net.Dialer
+	conn, err := dialer.DialContext(ctxTimeout, "tcp", p.Addr())
+	if err != nil {
+		return
+	}
+	p.AssociateConnection(conn)
+	defer p.Disconnect()
+
+	// Wait for the verack message or timeout in case of failure.
+	select {
+	case <-verack:
+		// Mark this peer as a good node.
+		amgr.Good(p.NA().IP, p.Services(), p.ProtocolVersion())
+
+		// Ask peer for some addresses.
+		p.QueueMessage(wire.NewMsgGetAddr(), nil)
+
+	case <-time.After(defaultNodeTimeout):
+		log.Printf("verack timeout on peer %v", p.Addr())
+		return
+	case <-ctx.Done():
+		return
+	}
+
+	select {
+	case <-onaddr:
+	case <-time.After(defaultNodeTimeout):
+		log.Printf("getaddr timeout on peer %v", p.Addr())
+	case <-ctx.Done():
+	}
+}
+
+func creep(ctx context.Context, netParams *chaincfg.Params) {
 	for {
+		if ctx.Err() != nil {
```

Comment:
Why use `ctx.Err()` here? In the rest of the dcrseeder code (and most other decred code afaik) the pattern used for infinite loop with an exit is thus:

```
select {
    case <-ctx.Done():
        // Exit
    default:
       // Do work
}
```

Reply:
Just a touch more concise, and no less appropriate given there's no other select.  This `Err` pattern is fairly common in dcrwallet, but TBH I don't much care which way we use here.